pkg/parse/updater.go

```diff
@@ -115,54 +133,119 @@ func (u *updater) Update(ctx context.Context, cache *cacheForCommit) status.Mult
 		u.updateMux.Unlock()
 	}()
 
-	objs := filesystem.AsCoreObjects(cache.objsToApply)
-
-	// Update the declared resources so that the Remediator immediately
-	// starts enforcing the updated state.
-	if !cache.resourceDeclSetUpdated {
-		var validationErrs status.MultiError
-		objs, validationErrs = u.resources.Update(ctx, objs)
-		u.setValidationErrs(validationErrs)
-		if validationErrs != nil {
-			klog.Warningf("Failed to validate declared resources: %v", validationErrs)
-			return validationErrs
-		}
+	updateErrs := u.update(ctx, cache)
+
+	// Prepend current conflict errors
+	var errs status.MultiError
+	errs = status.Append(errs, u.conflictErrors())
+	errs = status.Append(errs, updateErrs)
+	return errs
+}
 
+// update performs most of the work for `Update`, making it easier to
+// consistently prepend the conflict errors.
+func (u *updater) update(ctx context.Context, cache *cacheForCommit) status.MultiError {
+	// Stop remediator workers.
+	// This prevents objects been updated in the wrong order (dependencies).
+	// Continue watching previously declared objects and updating the queue.
+	// Queued objects will be remediated when the workers are started again.
+	u.remediator.Pause()
+
+	// Update the declared resources (source of truth for the Remediator)
+	var objs []client.Object
+	if cache.resourceDeclSetUpdated {
+		// No previous validation errors.
+		// Use previously validated and sanitized unstructured objects.
+		objs = u.resources.DeclaredObjects()
+	} else {
+		objs = filesystem.AsCoreObjects(cache.objsToApply)
+		var err status.MultiError
+		objs, err = u.declare(ctx, objs)
+		if err != nil {
+			return err
+		}
+		// Only mark the declared resources as updated if there were no (non-blocking) parse errors.
+		// This ensures the update will be retried.
 		if cache.parserErrs == nil {
 			cache.resourceDeclSetUpdated = true
 		}
 	}
-	// else - there were no previous validation errors
 
 	var gvks map[schema.GroupVersionKind]struct{}
-	var applyErrs status.MultiError
 	if cache.hasApplierResult {
+		// No previous apply errors.
+		// Use previously applied GVK set.
 		gvks = cache.applierResult
-		// there were no previous applier errors
 	} else {
-		applyStart := time.Now()
-		gvks, applyErrs = u.applier.Apply(ctx, objs)
-		metrics.RecordApplyDuration(ctx, metrics.StatusTagKey(applyErrs), cache.source.commit, applyStart)
-		if applyErrs != nil {
-			klog.Warningf("Failed to apply declared resources: %v", applyErrs)
-		} else if cache.parserErrs == nil {
+		var err status.MultiError
+		gvks, err = u.apply(ctx, objs, cache.source.commit)
+		if err != nil {
+			return err
```

Comment:
This is a behavioral change. Before this PR, it continues to update the remediator's watches even if there're apply errors.
Do we want to stop updating the watches when there're apply errors, especially for the non-blocking errors (missing CRD)?

Reply:
Should we update the watches before applying?
The declared resources are already updated beforehand.
And we're effectively disregarding any drift that happens to objects intended for deletion.
So it doesn't really matter if we stop watching them before they're deleted.

Reply:
The watches are always updated after applying.
My question is about whether we should update the watches when there're apply errors?

After checking the `apply` function, if error is not nil, the gvks will be nil. It basically won't update any watches anyway.